### PR TITLE
Add vtk_poly_data_reader [vtk_to_cmesh 4/6]

### PR DESCRIPTION
**Pull request by Davknapp**
First implementation of a reader for vtkPolyData.

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

- [x] The author added a BSD statement to `doc/` (or already has one)
- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

- [x] All tests pass (in various configurations, this should be executed automatically in a github action)
- [x] New source/header files are properly added to the Makefiles
- [ ] New Datatypes are added to t8indent_custom_datatypes.txt
- [ ] The reviewer executed the new code features at least once and checked the results manually
- [ ] The code is covered in an existing or new test case
- [x] New tests use the Google Test framework
- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [x] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

### CODE REVIEW

File: src/Makefile.am [hunk 1]
```diff
@@ -74,9 +74,10 @@ libt8_internal_headers = \
   src/t8_cmesh/t8_cmesh_types.h src/t8_cmesh/t8_cmesh_partition.h \
   src/t8_cmesh/t8_cmesh_refine.h src/t8_cmesh/t8_cmesh_copy.h \
   src/t8_cmesh/t8_cmesh_offset.h \
+  src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx \
   src/t8_forest/t8_forest_cxx.h  \
   src/t8_forest/t8_forest_ghost.h \
-  src/t8_forest/t8_forest_balance.h src/t8_forest/t8_forest_types.h \
+  src/t8_forest/t8_forest_balance.h src/t8_forest/t8_forest_types.h \
```
Comment (sandro-elsweijer) [inline src/Makefile.am:80]:
```suggestion
  src/t8_forest/t8_forest_balance.h src/t8_forest/t8_forest_types.h \
```

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+#include <vtkOBJReader.h>
+#include <vtkPLYReader.h>
+#include <vtkPolyDataReader.h>
+#include <vtkSTLReader.h>
+#include <vtkXMLPolyDataReader.h>
+
+void
+t8_read_poly (const char *filename, vtkSmartPointer < vtkPolyData > grid)
+{
+  char                tmp[BUFSIZ], *extension;
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:36]:
```suggestion
  char                tmp[BUFSIZ];
  char               *extension;
```

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  /* Get the file-extension to decide which reader to use. */
+  strcpy (tmp, filename);
+  extension = strtok (tmp, ".");
+  extension = strtok (NULL, ".");
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:40]: Why are you overwriting your pointer with a search in `NULL`?

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    char                tmp[BUFSIZ], *extension;
+    /* Get the file-extension to decide which reader to use. */
+    strcpy (tmp, filename);
+    extension = strtok (tmp, ".");
+    extension = strtok (NULL, ".");
+    T8_ASSERT (strcmp (extension, ""));
+
+    /* Read the file depending on the extension. Not all readers have
+     * a built-in check if the file is readable.  */
+    if (strcmp (extension, "ply") == 0) {
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:61]: The extensions were already checked.
You already checked, that the extension is `"vtp"`, it cannot be `"ply"` here.
Did you accidentally copy parts of the function into this if-clause?

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]
This is only a partial review, because the code seems broken

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx (hunk 3 above)
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:40]: I am not. Calling strtok twice, once with a string, once with NULL calls it again with the result of the first call.

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx (hunk 4 above)
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:61]: Yes, I updated it. Thanks for catching it.

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx (hunk 3 above)
Comment (sandro-elsweijer) [reply src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:41]: Maybe you want to search for the last point in the filename. This way, you can also catch filenames with multiple points.

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  char                tmp[BUFSIZ];
+  char               *extension;
+  /* Get the file-extension to decide which reader to use. */
+  strcpy (tmp, filename);
+  extension = strtok (tmp, ".");
+  extension = strtok (NULL, ".");
+  T8_ASSERT (strcmp (extension, ""));
+
+  /* Read the file depending on the extension. Not all readers have
+   * a built-in check if the file is readable. */
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:45]: If not all readers can check, if a file is readable, you can check if the file at least exists. This way, you can already handle one of the most common user mistakes.

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    reader->SetFileName (filename);
+    reader->Update ();
+    grid->ShallowCopy (reader->GetOutput ());
+    return;
+  }
+  else if (strcmp (extension, "vtp") == 0) {
+    vtkNew < vtkXMLPolyDataReader > reader;
+    reader->SetFileName (filename);
+    if (!reader->CanReadFile (filename)) {
+      t8_errorf ("Unable to read file.\n");
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:57]: Do you want to include the filename in the error message? Maybe the user wants to read multiple files and wants to know which exact file is unreadable

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    grid->ShallowCopy (reader->GetOutput ());
+    return;
+  }
+  else if (strcmp (extension, "vtk") == 0) {
+    vtkNew < vtkPolyDataReader > reader;
+    reader->SetFileName (filename);
+    reader->Update ();
+    if (!reader->IsFilePolyData ()) {
+      t8_errorf
+        ("File-content is not polydata. If it is a vtkUnstructuredGrid use the unstructured Grid reader.");
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:84]: How do I, as a user, use the unstructured Grid reader, if it is not implemented in the vtk reader?

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    vtkNew < vtkBYUReader > reader;
+    reader->SetGeometryFileName (filename);
+    reader->Update ();
+    grid->ShallowCopy (reader->GetOutput ());
+    return;
+  }
+  else {
+    /* Return NULL if the reader is not used correctly. */
+    t8_global_errorf ("Please use .ply, .vtp, .obj, .stl, .vtk or .g file\n");
+    return NULL;
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:100]: In other functions we return an `int` to signal, if a function call was successful or not

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  else if (strcmp (extension, "g") == 0) {
+    vtkNew < vtkBYUReader > reader;
+    reader->SetGeometryFileName (filename);
+    reader->Update ();
+    grid->ShallowCopy (reader->GetOutput ());
+    return;
+  }
+  else {
+    /* Return NULL if the reader is not used correctly. */
+    t8_global_errorf ("Please use .ply, .vtp, .obj, .stl, .vtk or .g file\n");
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:99]: Same as above: You could include the file name so the user knows which exact file is meant

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx (hunk 7 above)
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:93]: It is implemented on another branch

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with t8code; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+#if T8_WITH_VTK
+#include <t8_cmesh_vtk_polydata.hxx>
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:24]:
```suggestion
#include <t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx>
```

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+#include <vtkPolyData.h>
+#include <vtkBYUReader.h>
+#include <vtkOBJReader.h>
+#include <vtkPLYReader.h>
+#include <vtkPolyDataReader.h>
+#include <vtkSTLReader.h>
+#include <vtkXMLPolyDataReader.h>
+
+int
+t8_read_poly (const char *filename, vtkSmartPointer < vtkPolyData > grid)
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:34]:
```bash
In file included from ./../t8code/src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:25:
./../t8code/src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx:50:53: error: 'vtkPolyData' was not declared in this scope; did you mean 'vtkPolyData_h'?
   50 |                                   vtkSmartPointer < vtkPolyData > grid);
      |                                                     ^~~~~~~~~~~
      |                                                     vtkPolyData_h
./../t8code/src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx:50:65: error: template argument 1 is invalid
   50 |                                   vtkSmartPointer < vtkPolyData > grid);
```

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+{
+  char                tmp[BUFSIZ];
+  char               *extension;
+  /* Get the file-extension to decide which reader to use. */
+  strcpy (tmp, filename);
+  extension = strrchr (tmp, ".") + 1;
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:40]: which compiler are you using?

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx (hunk 11 above)
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:34]: Which vtk version are you using, and have you linked it properly?

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+*/
+
+/**
+ * This file contains all helper functions to translate vtk-polydata. 
+ * 
+ */
+
+#ifndef T8_CMESH_VTK_POLYDATA
+#define T8_CMESH_VTK_POLYDATA
+
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx:30]: That is odd, I am able to compile it.

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx (hunk 10 above)
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:24]: Why should this be necessary? My compiler finds the headers just fine.

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx (hunk 13 above)
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx:30]: Our CI is able to compile it in various configurations, too.

**Comment by sandro-elsweijer** [comment]
I found the issue. Due to not including `t8.h`, `t8_config.h` is not included and in `t8_config.h` the `T8_WITH_VTK` flag is set. Therefore, your code does not get compiled, even though you linked against VTK. The copy/paste error from the first review should at least have stopped your compiler.

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx (hunk 10 above)
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:24]:
```suggestion
#include <t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx>
#if T8_WITH_VTK
```
If we include `t8.h` in `t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx`, then we have to include it before the `T8_WITH_VTK`, so the flag is set before the compiler skips this code

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    return 0;
+  }
+  else {
+    /* Return NULL if the reader is not used correctly. */
+    t8_global_errorf
+      ("Please use .ply, .vtp, .obj, .stl, .vtk or .g file %s\n", filename);
+    return 1;
+  }
+}
+#endif
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.cxx:113]:
```suggestion
#endif /* T8_WITH_VTK */
```

File: src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+ * \param[in] filename  The name of the file
+ * \param[in, out] grid On input a vtkSmartPointer, that will hold the grid described in
+ *                      \a filename.
+ * \return              0 if the file was read successfully, non-zero otherwise.               
+ * 
+ */
+int                 t8_read_poly (const char *filename,
+                                  vtkSmartPointer < vtkPolyData > grid);
+
+#endif
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_vtk_to_t8/t8_cmesh_vtk_polydata.hxx:51]:
```suggestion
#endif /* T8_WITH_VTK */
```

**Review by sandro-elsweijer — APPROVED** [review]
Last two comments :)

**Pull request merged**